tests/test_api.py

```diff
@@ -55,3 +56,49 @@ def test_passthrough(self):
         """Test instances are passed through unmodified."""
         a = A(name='charlie')
         self.assertEqual(a, self.resolver.make(a))
+
+
+def _dummy_training_function(config):
+    """A dummy training function without actual training."""
+    # instantiate from configuration
+    to_resolve = config["choice"]
+    if isinstance(to_resolve, dict):
+        query = to_resolve.pop("query")
+        kwargs = to_resolve
+    else:
+        query = to_resolve
+        kwargs = None
+    resolver = Resolver([A, B, C], base=Base)
+    instance = resolver.make(query, pos_kwargs=kwargs)
+    if instance.name == "charlie":
+        mean_loss = 1.0
+    else:
+        mean_loss = 2.0
+
+    import ray.tune
+    ray.tune.report(mean_loss=mean_loss)
+
+
+def test_ray():
```

Comment:
this example has a lot of code duplication - why don't you put it inside a unittest TestCase so you can save the resolver as a state variable and make dummy_training_function also a function of the class

Reply:
With this modification, I got the following error

```python-traceback
Error
Traceback (most recent call last):
  File "/usr/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/usr/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "../class-resolver/tests/test_api.py", line 64, in test_ray
    analysis = ray.tune.run(
  File "../venv/class-resolver/lib/python3.8/site-packages/ray/tune/tune.py", line 306, in run
    experiments[i] = Experiment(
  File "../venv/class-resolver/lib/python3.8/site-packages/ray/tune/experiment.py", line 149, in __init__
    self._run_identifier = Experiment.register_if_needed(run)
  File "../venv/class-resolver/lib/python3.8/site-packages/ray/tune/experiment.py", line 301, in register_if_needed
    raise type(e)(str(e) + " " + extra_msg) from None
_pickle.PicklingError: ray.cloudpickle.dumps(<class 'ray.tune.function_runner.wrap_function.<locals>.ImplicitFunc'>) failed.
To check which non-serializable variables are captured in scope, re-run the ray script with 'RAY_PICKLE_VERBOSE_DEBUG=1'. Other options: 
-Try reproducing the issue by calling `pickle.dumps(trainable)`. 
-If the error is typing-related, try removing the type annotations and try again.
```

Reply:
I suppose it is connected to ray, which creates new processes for the experiments.

Reply:
With [ray.tune.with_parameters](https://docs.ray.io/en/master/tune/api_docs/trainable.html?highlight=with_parameters#ray.tune.with_parameters) and a static method it works, except that I get some redis warnings now :sweat_smile: 
```
../venv/class-resolver/lib/python3.8/site-packages/ray/node.py:827: ResourceWarning: unclosed file <_io.TextIOWrapper name='/tmp/ray/session_2021-03-31_12-29-09_243337_20619/logs/redis-shard_0.err' mode='a' encoding='utf-8'>
  self.start_redis()
ResourceWarning: Enable tracemalloc to get the object allocation traceback
../venv/class-resolver/lib/python3.8/site-packages/ray/node.py:827: ResourceWarning: unclosed file <_io.TextIOWrapper name='/tmp/ray/session_2021-03-31_12-29-09_243337_20619/logs/redis-shard_0.out' mode='a' encoding='utf-8'>
  self.start_redis()
```